src/HexagonOptimize.cpp

```diff
@@ -1079,11 +1081,64 @@ class OptimizePatterns : public IRMutator {
             // that they generate.
             internal_assert(op->args.size() == 3);
             return mutate(lower_lerp(op->args[0], op->args[1], op->args[2]));
+        } else if (op->is_intrinsic(Call::div_round_to_zero) &&
+                   !op->type.is_float() && op->type.is_vector()) {
+            internal_assert(op->args.size() == 2);
+            Expr a = op->args[0];
+            Expr b = op->args[1];
+            // Run bounds analysis to estimate the range of result.
+            Expr extent_upper = find_constant_bound(a / b, Direction::Upper, bounds);
+            const uint64_t *upper_bound = as_const_uint(extent_upper);
+
+            a = mutate(a);
+            b = mutate(b);
+            debug(1) << "Using long div: (num: " << a << "); (den: " << b << ")\n";
+            if (a.type().is_uint()) {
+                return unsigned_long_div(a, b, upper_bound);
+            }
+            // Use unsigned long div for signed integer division.
+            Type ty = a.type().with_code(Type::UInt);
+            Expr a_unsigned = cast(ty, abs(a));
+            Expr b_unsigned = cast(ty, abs(b));
+            Expr q = cast(a.type(),
+                          unsigned_long_div(a_unsigned, b_unsigned, upper_bound));
+            // Check the sign required for the quotient.
+            Expr a_neg = a >> make_const(a.type(), (a.type().bits() - 1));
+            Expr b_neg = b >> make_const(a.type(), (a.type().bits() - 1));
+            return q * ((a_neg ^ b_neg) | 1);
         } else {
             return IRMutator::visit(op);
         }
     }
 
+    template<typename NodeType, typename T>
+    NodeType visit_let(const T *op) {
+        // We only care about vector lets.
+        if (op->value.type().is_vector()) {
+            bounds.push(op->name, bounds_of_expr_in_scope(op->value, bounds));
+        }
+        NodeType node = IRMutator::visit(op);
+        if (op->value.type().is_vector()) {
+            bounds.pop(op->name);
+        }
+        return node;
+    }
+
+    Expr visit(const Let *op) override {
+        return visit_let<Expr>(op);
+    }
+
+    Stmt visit(const LetStmt *op) override {
+        return visit_let<Stmt>(op);
+    }
+
+    Expr visit(const Div *op) override {
+        if (!op->type.is_float() && op->type.is_vector()) {
+            return mutate(lower_int_uint_div(op->a, op->b));
```

Comment:
Doesn't this need to handle signed integers? I think it should call a helper that includes the logic in the `div_round_to_zero` handler.

Also, I think this is going to have to work with `lower_euclidean_div` somehow, in order to get the intended behavior. Or we need the reverse logic for `div_round_to_zero`.

Reply:
lower_int_uint_div can handle signed as well as unsigned integers. So we don't need to do anything else. Sorry I don't understand the need for the helper function here. It would be helpful if you can provide some more details on it.

Right. This would work with lower_euclidean_div. In cases other than lower_euclidean_div we change the division to mul with shift which is better than long_div. So I guess we only want to use long_div if the lower_int_uint_div generates calls to div_round_to_zero.